nixos/lib/systemd-lib.nix

```diff
@@ -11,6 +11,9 @@ in rec {
 
   mkPathSafeName = lib.replaceChars ["@" ":" "\\" "[" "]"] ["-" "-" "-" "" ""];
 
+  # a type for options that take a unit name
+  unitNameType = types.strMatching "[a-zA-Z0-9@%:_.\\-]+[.](service|socket|device|mount|automount|swap|target|path|timer|scope|slice)";
```

Comment:
How did you come up with the character class?

Reply:
man systemd.unit

>       Valid unit names consist of a "name prefix" and a dot and a suffix specifying the unit
>        type. The "unit prefix" must consist of one or more valid characters (ASCII letters,
>        digits, ":", "-", "_", ".", and "\"). The total length of the unit name including the
>        suffix must not exceed 256 characters. The type suffix must be one of ".service",
>        ".socket", ".device", ".mount", ".automount", ".swap", ".target", ".path", ".timer",
>        ".slice", or ".scope".
> 
>        Units names can be parameterized by a single argument called the "instance name". The unit
>        is then constructed based on a "template file" which serves as the definition of multiple
>        services or other units. A template unit must have a single "@" at the end of the name
>        (right before the type suffix). The name of the full unit is formed by inserting the
>        instance name between "@" and the unit type suffix. In the unit file itself, the instance
>        parameter may be referred to using "%i" and other specifiers, see below.
>